crates/rpc-client/src/withdrawal.rs

```diff
@@ -33,41 +56,85 @@ fn verify_l1_sudt_script(info: &CellInfo, l1_sudt_script_hash: &Byte32) -> Resul
     Ok(())
 }
 
-fn verify_finalized_owner_lock(info: &CellInfo, last_finalized_block_number: u64) -> Result<()> {
-    let args: Bytes = info.output.lock().args().unpack();
+#[derive(Debug)]
+enum OwnerLock {
+    Owner(Script),
+    V1Deposit(Script),
+}
 
-    let lock_args_end = 32 + WithdrawalLockArgs::TOTAL_SIZE;
-    let owner_lock_start = lock_args_end + 4; // u32 owner lock length
-    if args.len() <= owner_lock_start {
-        bail!("no owner lock");
-    }
+#[derive(Debug)]
+struct WithdrawalLock {
+    lock_args: WithdrawalLockArgs,
+    owner_lock: OwnerLock,
+}
 
-    let lock_args = match WithdrawalLockArgsReader::verify(&args.slice(32..lock_args_end), false) {
-        Ok(()) => WithdrawalLockArgs::new_unchecked(args.slice(32..lock_args_end)),
-        Err(_) => bail!("invalid withdrawal lock args"),
-    };
+impl WithdrawalLock {
+    fn from_cell(info: &CellInfo) -> Result<WithdrawalLock> {
+        let args: Bytes = info.output.lock().args().unpack();
 
-    if lock_args.withdrawal_block_number().unpack() > last_finalized_block_number {
-        bail!("unfinalized");
+        let lock_args_end = 32 + WithdrawalLockArgs::TOTAL_SIZE;
+        let owner_lock_start = lock_args_end + 4; // u32 owner lock length
+        let args_len = args.len();
+        if args_len <= owner_lock_start {
+            bail!("no owner lock");
+        }
+
+        let lock_args =
+            match WithdrawalLockArgsReader::verify(&args.slice(32..lock_args_end), false) {
+                Ok(()) => WithdrawalLockArgs::new_unchecked(args.slice(32..lock_args_end)),
+                Err(_) => bail!("invalid withdrawal lock args"),
+            };
+
+        let mut owner_lock_len_buf = [0u8; 4];
+        owner_lock_len_buf.copy_from_slice(&args.slice(lock_args_end..owner_lock_start));
+        let owner_lock_len = u32::from_be_bytes(owner_lock_len_buf) as usize;
```

Comment:
`u32::from_be_bytes(args[lock_args_end..owner_lock_start].try_into().unwrap())`?

(`args[lock_args_end..owner_lock_start]` COULD be a bit faster than `slice` because it does not need to create a new `Bytes` instance.)

Might also worth looking at `Buf` methods and see whether using them would make the code more clear, i.e. `get_u32`, `get_bytes`, etc: https://docs.rs/bytes/latest/bytes/buf/trait.Buf.html